src/auth/tenant.ts

```diff
@@ -0,0 +1,199 @@
+/*!
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as utils from '../utils';
+import * as validator from '../utils/validator';
+import {AuthClientErrorCode, FirebaseAuthError} from '../utils/error';
+import {
+  EmailSignInConfig, EmailSignInConfigServerRequest, EmailSignInProviderConfig,
+} from './auth-config';
+
+/** The server side tenant type enum. */
+export type TenantServerType = 'LIGHTWEIGHT' | 'FULL_SERVICE' | 'TYPE_UNSPECIFIED';
+
+/** The client side tenant type enum. */
+export type TenantType = 'lightweight' | 'full_service' | 'type_unspecified';
+
+/** The TenantOptions interface used for create/read/update tenant operations. */
+export interface TenantOptions {
+  displayName?: string;
+  type?: TenantType;
+  emailSignInConfig?: EmailSignInProviderConfig;
+}
+
+/** The corresponding server side representation of a TenantOptions object. */
+export interface TenantOptionsServerRequest extends EmailSignInConfigServerRequest {
+  displayName?: string;
+  type?: TenantServerType;
+}
+
+/** The tenant server response interface. */
+export interface TenantServerResponse {
+  name: string;
+  type?: TenantServerType;
+  displayName?: string;
+  allowPasswordSignup: boolean;
+  enableEmailLinkSignin: boolean;
+}
+
+/** The interface representing the listTenant API response. */
+export interface ListTenantsResult {
+  tenants: Tenant[];
+  pageToken?: string;
+}
+
+
+/**
+ * Tenant class that defines a Firebase Auth tenant.
+ */
+export class Tenant {
+  public readonly tenantId: string;
+  public readonly type?: TenantType;
+  public readonly displayName?: string;
+  public readonly emailSignInConfig?: EmailSignInConfig | null;
+
+  /**
+   * Builds the corresponding server request for a TenantOptions object.
+   *
+   * @param {TenantOptions} tenantOptions The properties to convert to a server request.
+   * @param {boolean} createRequest Whether this is a create request.
+   * @return {object} The equivalent server request.
+   */
+  public static buildServerRequest(
```

Comment:
Just a suggestion (feel free to ignore). I think we can make the usage of this method more readable if we made this method private, and added 2 additional public methods:

```
public static buildCreateRequest(tenantOptions: TenantOptions) {
  return buildServerRequest(tenantOptions, true);
}

public static buildUpdateRequest(tenantOptions: TenantOptions) {
  return buildServerRequest(tenantOptions, false);
}
```